mux/mux.go

```diff
@@ -3,104 +3,87 @@ package mux
 import (
 	"context"
 	"errors"
-	"net"
+	"log"
 	"net/http"
+	"strings"
 	"time"
 
-	"github.com/soheilhy/cmux"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
 	"google.golang.org/grpc"
 )
 
 const defaultGracePeriod = 5 * time.Second
+const grpcRequestMarker = "application/grpc"
 
-// Serve starts a TCP listener and serves the registered protocol
-// servers (i.e., http, grpc, etc.) and blocks until server exits.
-// Context can be cancelled to perform graceful shutdown.
-// Use WithHTTP() and WithGRPC() to register procol servers.
+// Serve starts a TCP listener and serves the registered protocol servers and blocks
+// until server exits. Context can be cancelled to perform graceful shutdown.
+// Use WithHTTP() and WithGRPC() to register protocol servers.
 func Serve(ctx context.Context, listenAddr string, opts ...Option) error {
-	var mux cmuxWrapper
+	var mux protocolMuxServer
 	for _, opt := range opts {
 		if err := opt(&mux); err != nil {
 			return err
 		}
 	}
-	if mux.httpServer == nil && mux.grpcServer == nil {
+	if mux.httpHandler == nil && mux.grpcServer == nil {
 		return errors.New("at-least one of http & grpc server must be set")
 	}
-
 	return mux.Serve(ctx, listenAddr)
 }
 
-type cmuxWrapper struct {
-	httpServer  *http.Server
+type protocolMuxServer struct {
```

Comment:
aren't we multiplexing `grpc` & `http`? its not a protocol we are multiplexing, but just another way to communicate over http 1 or 2.
`protocolMuxServer` is bit misleading there, as we are just doing grpc vs http here, and not multiplexing protocols like `ftp`, `http`, `smtp` .. etc using this

Reply:
i was thinking of "protocol" in a more general term (= bunch of rules defined together to facilitate a specific way of communication.. like json-rpc protocol)..

but i get how it would be confusing when you are thinking of net protocols specifically.. we can change it. any suggestions? we could just call this `muxServer` also i guess.

Reply:
imo `muxServer` would be better
wdys @ravisuhag ?

Reply:
i have already renamed it to `muxServer`. will wait for other suggestions. if nothing else, will go with that name itself.

Reply:
`muxServer` sounds good.